roslibrust_codegen/src/lib.rs

```diff
@@ -0,0 +1,230 @@
+    message_files.extend_from_slice(&service_files[..]);
+    let messages = parse_message_files(message_files).unwrap();
+    let struct_definitions = messages
+        .into_iter()
+        .map(|(_, message)| generate_struct(message))
+        .collect::<Vec<TokenStream>>();
+
+    quote! {
+        #(#struct_definitions)*
+
+    }
+}
+
+struct MessageMetadata {
+    pub package: String,
+    pub path: PathBuf,
+    pub seen_count: i32,
+    pub parsed: MessageFile,
+    pub unparsed: String,
+}
+
+fn parse_message_files(
+    msg_paths: Vec<(String, PathBuf)>,
+) -> std::io::Result<BTreeMap<String, MessageFile>> {
+    let mut parsed_messages = VecDeque::new();
+    for (pkg, path) in msg_paths {
+        let contents = std::fs::read_to_string(&path)?;
+        let name = path.file_stem().unwrap().to_str().unwrap();
+        match path.extension().unwrap().to_str().unwrap() {
+            "srv" => {
+                parse_ros_service_file(contents, name.to_string(), &pkg)
+                    .into_iter()
+                    .for_each(|(parsed, unparsed)| {
+                        parsed_messages.push_back(MessageMetadata {
+                            package: pkg.clone(),
+                            path: path.clone(),
+                            seen_count: 0,
+                            parsed,
+                            unparsed,
+                        });
+                    });
+            }
+            "msg" => {
+                let msg = parse_ros_message_file(contents.clone(), name.to_string(), &pkg);
+                parsed_messages.push_back(MessageMetadata {
+                    package: pkg.clone(),
+                    path: path.clone(),
+                    seen_count: 0,
+                    parsed: msg,
+                    unparsed: contents,
+                });
+            }
+            _ => {
+                panic!("File extension not recognized as a ROS file: {path:?}");
+            }
+        }
+    }
+    Ok(resolve_message_dependencies(parsed_messages))
+}
+
+fn resolve_message_dependencies(
+    mut parsed_msgs: VecDeque<MessageMetadata>,
+) -> BTreeMap<String, MessageFile> {
+    const MAX_PARSE_ITER_LIMIT: i32 = 2048;
+    let mut message_map = BTreeMap::new();
+
+    while let Some(MessageMetadata {
+        package,
+        path,
+        seen_count,
+        parsed,
+        unparsed,
+    }) = parsed_msgs.pop_front()
+    {
+        if seen_count > MAX_PARSE_ITER_LIMIT {
+            parsed.fields.iter().for_each(|field| {
+                if ROS_TYPE_TO_RUST_TYPE_MAP.contains_key(field.field_type.field_type.as_str()) {
+                    println!("RESOLVED PRIMITIVE: {}", field.field_type.field_type);
+                } else if message_map.contains_key(&field.field_type.field_type) {
+                    println!("RESOLVED MSG: {}", field.field_type.field_type);
+                } else {
+                    println!("UNRESOLVED: {}", field.field_type.field_type);
+                }
```

Comment:
eprintln! ?

Probably want a "MAX_PARSE_ITER_LIMIT Exceeded, likely circular dependency message?"

Reply:
Ah I actually intended to take this out. There's a message directly below for that.